Speed up tensor.get_device(), is_cuda(), is_sparse() by avoiding dispatches

zou3519 opened the pull request:
`tensor.get_device()` went through two dispatches: once to the native
function
`get_device()`, and another when `get_device` calls `_th_get_device()`.
This PR avoids the dispatch by directly implementing the `get_device`
function
as a method on Tensor.

Future Work:
- Investigate caching Device on TensorImpl. This will probably bring the
  tensor.get_device down to 2ns, but I'm not sure it's worth it.

before:
```
------------------------------------------------------------------------
Benchmark                                 Time           CPU Iterations
------------------------------------------------------------------------
BM_TensorTypeId                           0 ns          0 ns 1000000000
BM_TensorType                             8 ns          8 ns   89407911
BM_TensorIsCuda                          24 ns         24 ns   29313017
BM_TensorIsSparse                        27 ns         27 ns   26083160
BM_TensorTypeIsCuda                      11 ns         11 ns   65128120
BM_TensorNumel                           11 ns         11 ns   68314492
BM_TensorGetDevice                       71 ns         71 ns    9633125
BM_DeviceGuardCtor                      173 ns        173 ns    4067173
BM_DeviceGuard                          232 ns        232 ns    3009690
```

after:
```
------------------------------------------------------------------------
Benchmark                                 Time           CPU Iterations
------------------------------------------------------------------------
BM_TensorTypeId                           0 ns          0 ns 1000000000
BM_TensorType                            10 ns         10 ns   69803872
BM_TensorIsCuda                           2 ns          2 ns  321626683
BM_TensorIsSparse                         6 ns          6 ns  177045382
BM_TensorNumel                           12 ns         12 ns   58770533
BM_TensorGetDevice                        4 ns          4 ns  128113396
BM_DeviceGuardCtor                       52 ns         52 ns   14997278
BM_DeviceGuard                          158 ns        158 ns    5767248

```

zou3519 committed:
Speed up tensor.get_device() by avoiding dispatches

`tensor.get_device()` went through two dispatches: once to the native
function
`get_device()`, and another when `get_device` calls `_th_get_device()`.
This PR avoids the dispatch by directly implementing the `get_device`
function
as a method on Tensor.

Future Work:
- Investigate caching Device on TensorImpl. This will probably bring the
  tensor.get_device down to 10ns, but I'm not sure it is worth it.

```
  after:
  ------------------------------------------------------------------------
  Benchmark                                 Time           CPU
  Iterations
  ------------------------------------------------------------------------
  BM_TensorTypeId                           0 ns          0 ns
  1000000000
  BM_TensorType                            10 ns         10 ns
  69803872
  BM_TensorIsCuda                           2 ns          2 ns
  321626683
  BM_TensorIsSparse                         6 ns          6 ns
  177045382
  BM_TensorNumel                           12 ns         12 ns
  58770533
  BM_TensorGetDevice                        4 ns          4 ns
  128113396
  BM_DeviceGuardCtor                       52 ns         52 ns
  14997278
  BM_DeviceGuard                          158 ns        158 ns
  5767248

  before:

  ------------------------------------------------------------------------
  Benchmark                                 Time           CPU
  Iterations
  ------------------------------------------------------------------------
  BM_TensorTypeId                           0 ns          0 ns
  1000000000
  BM_TensorType                             8 ns          8 ns
  89407911
  BM_TensorIsCuda                          24 ns         24 ns
  29313017
  BM_TensorIsSparse                        27 ns         27 ns
  26083160
  BM_TensorTypeIsCuda                      11 ns         11 ns
  65128120
  BM_TensorNumel                           11 ns         11 ns
  68314492
  BM_TensorGetDevice                       71 ns         71 ns
  9633125
  BM_DeviceGuardCtor                      173 ns        173 ns
  4067173
  BM_DeviceGuard                          232 ns        232 ns
  3009690
 ```

ezyang commented on aten/src/ATen/core/TensorMethods.h:
```diff
@@ -1240,6 +1237,21 @@ inline Device Tensor::device() const {
+  }
+  // TODO(rzou): Investigate caching device on TensorImpl.
+  // It'll probably be faster than having this indirection through Storage
+  return impl_->storage().unsafeGetStorageImpl()->device().index();
```
Err... I'm pretty sure you don't need `unsafeGetStorageImpl()` here; we should give you `device()` directly on Storage.

ezyang commented on aten/src/ATen/core/TensorMethods.h:
```diff
@@ -1240,6 +1237,21 @@ inline Device Tensor::device() const {
+  // to a native function.
+  const auto& mytype = type();
+  if (!mytype.is_cuda()) {
+    return -1;
```
Didn't it use to error in this case?

zou3519 replied:
Yes, thank you for the catch!

ezyang reviewed (approved):
OK, but I do have some comments. Please read.

ezyang commented:
Hmm... I suspect all of the pre-canned functions on Tensor have hand written bindings. Not great.

zou3519 committed:
Add manual python bindings for get_device

Copied & pasted them from the old generated/python_variable_methods &
generated/python_torch_functions.

zou3519 committed:
Address comments

zou3519 committed:
speed up get_device more

zou3519 committed:
Speed up is_cuda, is_sparse

zou3519 committed:
fix comment

ezyang commented on aten/src/ATen/core/TensorMethods.h:
```diff
@@ -1240,6 +1231,44 @@ inline Device Tensor::device() const {
+  return self.is_cuda();
+}
+
+inline bool Tensor::is_sparse() const {
```
This method would mean that sparse is no longer open, which could be problematic in practice because one could very reasonably want to add a `SparseHIPTensorId`, e.g.

If we are going to allow people to write code like this, it seems like a strong argument to make TensorTypeId as a closed enum, and then implement all of these "attribute" tests in the file (so we remember to update them when we add other TensorTypeIds).

If we *don't* want to allow this, then we had better have another, non-vdispatched way to check what the memory layout of a tensor is. Maybe some sort of "Layout" field in TensorImpl?

gchanan replied:
I think we should just dodge this question and force the implementation into TensorTypeId.  I don't have a strong preference for returning a bool or a layout enum or whatever for now.  But this isn't maintainable.

ezyang replied:
OK. @zou3519, just to make the ask clear: we'd like this implementation to live in `TensorImpl`, rather than as a method on `Tensor`. API doesn't have to change; we just want this to live closer to where we could make implementation changes to change how we implement this.

zou3519 replied:
Sounds good, thanks for the clarification!

zou3519 replied:
Do you want all of `{ is_sparse, is_cuda, get_device }` to live in `TensorImpl`?

zou3519 replied:
Also, @ezyang, @gchanan, is it still okay to special case these functions in `Tensor`? They would end up calling the relevant methods on TensorImpl.

I'm trying to avoid a type() dispatch because those are costly (8ns) compared to how fast `is_cuda` and `is_sparse` can theoretically be.

ezyang replied:
It's OK to special case them on tensor; just have them call the method on TensorImpl.

ezyang commented:
@zou3519 Have you given any thought on how to prevent people from virtualizing this method again? :)

zou3519 commented:
@ezyang Not sure. I imagine setting up a benchmark test for these functions would do the trick (virtualizing them vs not virtualizing is like a 10x difference), or just by adding a comment into the implementation

ezyang commented:
I guess a comment will do for now.

zou3519 commented:
https://stackoverflow.com/questions/22911112/how-to-detect-if-a-method-is-virtual suggests a way to detect if a method is virtual on gcc >= 7 and most if not all versions of clang. I'll try to see if I can do a static assert for this.

zou3519 committed:
Additional cleanup; address comments

zou3519 committed:
Fix Variable get_device

zou3519 commented:
Nvm, that stackoverflow post explains how to detect if a class has virtual methods, not whether if a specific method is virtual. I'll stick with the comment then -- I've moved the implementations to TensorImpl as requested.

zou3519 commented:
Test failures look unrelated (they are all build timeouts). 

When you get the chance, could you take another look at the changes again, @ezyang?

ezyang commented on aten/src/ATen/core/TensorImpl.h:
```diff
@@ -391,6 +413,13 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
   }
   bool compute_contiguous() const;
 
+  virtual int64_t get_device_slow() const {
```
This deserves a comment about why it exists, and why it's slow.

ezyang commented on aten/src/ATen/core/TensorImpl.h:
```diff
@@ -229,6 +229,28 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
     return is_contiguous_;
   }
 
+  bool is_sparse() const {
```
Footgun note: if you have a Variable, it will NEVER report true for any of these methods.

ezyang replied:
Potential footgun note: at the moment, variables have the same TensorTypeId as their corresponding tensor, but if this ever changes, you have the potential to be very sad. Not actionable at the moment, I think.

ezyang commented on aten/src/ATen/core/TensorImpl.h:
```diff
@@ -229,6 +229,28 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
 
+  bool is_sparse() const {
+    // NB: This method is not virtual and avoid dispatches for performance reasons.
+    const auto& tid = type_id();
```
This is harmless, in the sense that the temporary will get its scope extended for the reference, but it's totally unnecessary, because TensorTypeId is a value object and should be treated as such.

ezyang reviewed (approved):
This looks way better. Some minor comments.

zou3519 committed:
address comments: add comments

zou3519 committed:
Fix rebase
